src/analysis/retail/evoker/shared/modules/talents/hero/chronowarden/Primacy.tsx

```diff
@@ -0,0 +1,44 @@
+import Analyzer, { Options } from 'parser/core/Analyzer';
+import Statistic from 'parser/ui/Statistic';
+import STATISTIC_CATEGORY from 'parser/ui/STATISTIC_CATEGORY';
+import STATISTIC_ORDER from 'parser/ui/STATISTIC_ORDER';
+import { TALENTS_EVOKER } from 'common/TALENTS';
+import TalentSpellText from 'parser/ui/TalentSpellText';
+import SPELLS from 'common/SPELLS';
+import { HasteIcon, InformationIcon } from 'interface/icons';
+import { formatPercentage } from 'common/format';
+import { PRIMACY_HASTE_PER_STACK } from 'analysis/retail/evoker/shared/constants';
+
+class Primacy extends Analyzer {
+  constructor(options: Options) {
+    super(options);
+    this.active = this.selectedCombatant.hasTalent(TALENTS_EVOKER.PRIMACY_TALENT);
+  }
+
+  statistic() {
+    const buffUptime =
+      this.selectedCombatant.getBuffUptime(SPELLS.PRIMACY_BUFF.id) / this.owner.fightDuration;
+    const stackBuffUptimes = this.selectedCombatant.getStackBuffUptimes(SPELLS.PRIMACY_BUFF.id);
+    const hasteBuffPercentage =
+      (PRIMACY_HASTE_PER_STACK *
+        (stackBuffUptimes[1] + 2 * stackBuffUptimes[2] + 3 * stackBuffUptimes[3])) /
+      this.owner.fightDuration;
```

Comment:
Looking at this, wouldn't this result in a `NaN` value if any of the array entries are `undefined`? I'm not certain how often this would happen, but wouldn't hurt to check for it.
Using reduce like this could be a good solution. (This would also apply for MomentumShift module)
```suggestion
    const stackBuffUptimes = this.selectedCombatant.getStackBuffUptimes(SPELLS.PRIMACY_BUFF.id);
    const avgStackUptime = Object.values(stackBuffUptimes).reduce((acc,val,idx) =>  acc + (val * idx),0) // Not sure if this is proper name for this variable
    const hasteBuffPercentage =
      (PRIMACY_HASTE_PER_STACK *
        avgStackUptime) /
      this.owner.fightDuration;
```